File: Source/Samples/95_Periodic/code_adimmer.cpp
// APromptWindow application source code
#include <Urho3D/AUI/AUI.h>
#include <Urho3D/AUI/AEvents.h>
#include <Urho3D/AUI/APromptWindow.h>
#include <Urho3D/AUI/AMessageWindow.h>
#include <Urho3D/AUI/AFinderWindow.h>
#include <Urho3D/AUI/ADimmer.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_adimmer( AWidget *layout )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAdimmerEvent ));
}

void PeriodicApp::HandleModalCompleteEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace APromptComplete;

    AppLog( "Modal Selection event : the window " + eventData[P_TITLE].GetString()
            + " selection was " + eventData[P_SELECTED].GetString()
            + ", the button pressed was " + eventData[P_REASON].GetString());

    AUI* aui = GetSubsystem<AUI>();  // manually undim the app
    AWidget *someview = (AWidget *)aui->GetFocusedView();
    AWidget *dimmer = someview->FindWidget("dimmer");
    if ( dimmer )
        dimmer->Remove();

}

void PeriodicApp::HandleAdimmerEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    AWidget *someview = (AWidget *)widget->GetView();

    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uidimmercode" )
        {
            AppLog( "ADimmer support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_adimmer.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uidimmerlayout" )
        {
            AppLog( "ADimmer support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_adimmer.ui.txt", widget->GetParent() );
        }

        if (widget->GetId() == "modalfinder" )
        {
            AppLog( "ADimmer action : " + widget->GetId() + " was pressed ");
            APromptWindow *windowp = new APromptWindow(context_, someview, widget->GetId());
            SubscribeToEvent( windowp, E_APROMPTCOMPLETE, URHO3D_HANDLER(PeriodicApp, HandleModalCompleteEvent ));

            ADimmer *dimmer = new ADimmer(context_);  //manually set app dimming
            dimmer->SetId ("dimmer");
            someview->AddChild(dimmer);

            windowp->Show( "Modal Selection", "ADimmer in effect", "preset value", 0, 0, 0);
        }

        if (widget->GetId() ==  "modalok" )
        {
            AppLog( "ADimmer support : " + widget->GetId() + " was pressed ");
            AMessageWindow *mess2 = new AMessageWindow(context_, someview, "mymess2");
            mess2->Show("MessageWindow - OK", "Modal MessageWindow - OK button", UI_MESSAGEWINDOW_SETTINGS_OK, 1, 0, 0);
        }
        if (widget->GetId() ==  "modalokcancel" )
        {
            AppLog( "ADimmer support : " + widget->GetId() + " was pressed ");
            AMessageWindow *mess3 = new AMessageWindow(context_, someview, "mymess3");
            mess3->Show("MessageWindow - OK CANCEL", "Modal MessageWindow - OK CANCEL buttons", UI_MESSAGEWINDOW_SETTINGS_OK_CANCEL, 1, 0, 0);
        }
        if (widget->GetId() ==  "modalyesno" )
        {
            AppLog( "ADimmer support : " + widget->GetId() + " was pressed ");
            AMessageWindow *mess4 = new AMessageWindow(context_, someview, "mymess4");
            mess4->Show( "MessageWindow - YES NO", "this is a MessageWindow - YES NO buttons", UI_MESSAGEWINDOW_SETTINGS_YES_NO, 1, 0, 0);
        }
        if (widget->GetId() == "modalfilefinder" )
        {
            AppLog( "ADimmer action : " + widget->GetId() + " was pressed ");
            AFinderWindow *windowf = new AFinderWindow(context_, someview, widget->GetId());
            windowf->FindFile("Modal Find a File", "", 1, 0, 0);
        }
        if (widget->GetId() ==  "modalfolderfinder" )
        {
            AppLog( "ADimmer action : " + widget->GetId() + " was pressed ");
            AFinderWindow *windowd = new AFinderWindow(context_, someview, widget->GetId() );
            windowd->FindPath("Modal Find a Folder", "", 1, 0, 0);
        }

    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_adimmer.lua
-- ADimmer application source code


function setup_uidimmer( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUidimmerEvent")
    end
end

function HandleUidimmerEvent(eventType, eventData)
    local refid = eventData["RefID"]:GetString()
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return
    end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uidimmercode" then
            AppLog( "ADimmer support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_adimmer.lua", widget:GetParent())
        end
        if widget:GetId() ==  "uidimmerlayout" then
            AppLog( "ADimmer support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_adimmer.ui.txt", widget:GetParent())
        end
        if widget:GetId() == "modalfinder" then
            AppLog( "ADimmer action : " .. widget:GetId() .. " was pressed ")
            local windowp = APromptWindow:new( uiview, widget:GetId())
            windowp:Show( "Modal Prompt", "Modal Selection in window", "preset value", 1, 0, 0)
        end
        if widget:GetId() == "modalok" then
            AppLog( "ADimmer support : " .. widget:GetId() .. " was pressed ")
            local mess2 = AMessageWindow:new( uiview, "mymess2")
            mess2:Show("MessageWindow - OK", "Modal MessageWindow - OK button", UI_MESSAGEWINDOW_SETTINGS_OK, true, 0, 0)
        end
        if widget:GetId() == "modalokcancel" then
            AppLog( "ADimmer support : " .. widget:GetId() .. " was pressed ")
            local mess3 = AMessageWindow:new(uiview, "mymess3")
            SubscribeToEvent(mess3, "WidgetEvent", "HandleUimessagewindowEvent" )
            mess3:Show("MessageWindow - OK CANCEL", "Modal MessageWindow - OK CANCEL buttons", UI_MESSAGEWINDOW_SETTINGS_OK_CANCEL, true, 0, 0)
        end
        if widget:GetId() == "modalyesno" then
            AppLog( "ADimmer support : " .. widget:GetId() .. " was pressed ")
            local mess4 = AMessageWindow:new( uiview, "mymess4")
            mess4:Show( "MessageWindow - YES NO", "Modal MessageWindow - YES NO buttons", UI_MESSAGEWINDOW_SETTINGS_YES_NO, true, 0, 0)
        end

        if widget:GetId() == "modalfilefinder" then
            AppLog( "ADimmer action : " .. widget:GetId() .. " was pressed ")
            winf = AFinderWindow:new(uiview, widget:GetId())
            winf:FindFile("Find a File", "", 1, 0, 0)
        end
        if widget:GetId() ==  "modalfolderfinder" then
            AppLog( "ADimmer action : " .. widget:GetId() .. " was pressed ")
            windowd = AFinderWindow:new(uiview, widget:GetId())
            windowd:FindPath("Find a Folder", "", 1, 0, 0)
        end
    else
        AppLog( "ADimmer event : " .. widget:GetId() .. " event type = " .. eventData["Type"]:GetInt() )
   end
end

